### ISSUE TITLE
Is it possible to test middleware

### ISSUE THREAD

**Issue by newmanw** [underscopeio/sinon-mongoose#22]
Examples are stubbing out the entire mongoose model and seem to be preventing middleware from being called.  Is it possible to test mongoose middleware while mocking?

SomeSchema.pre('save', function(next) {
  // doing some pre save stuff here that I would like to test
});

**Comment by gaguirre** [comment]
Do you mean that mocking the `save` method will prevent the `SomeSchema.pre('save', ...)` hook to be called?

**Comment by newmanw** [comment]
@gaguirre yes as far as I can tell.   That being said I could be mocking the save wrong. ;) Is this something that should work if I am mocking save correctly?

**Comment by gaguirre** [comment]
Probably you are doing it well.

The thing is: if you're mocking the `save` method then it will not be called, so probably the `pre-save` hook is not called neither.
I'm wondering if it's something we should solve in `sinon-mongoose` or just provide a _recipe_ for this case.

@jniemin did you face this issue?

**Comment by newmanw** [comment]
@gaguirre I think solving in sinon-mongoose or a recipe are both valid options IMHO.

I use the mongoose middleware a lot and think its a great way to separate concerns.  That being said I have found it very hard to test that middleware.  I have tried to just mock the mongoose collection object directly, which in turn really just mocks the mongodb driver.  It works in a sense that all mongoose middleware is run when I do this.  However I am at the mercy of figuring out what type of object the mongodb driver returns back to mongoose so things still 'work'.

That being said, I am also looking at seeing if I can access the middleware 'hooks' in my tests directly.  That way I can test the middleware function directly rather than indirectly through a save/update/remove call.

**Comment by jniemin** [comment]
I didn't run in to it. But if you think about it, you should not need to test if middleware is called in your test. That is basically testing the mongoose logic, not your code. What I would do instead is test the function passed to middleware

```
var saveFunction = function(next){
  next();
}

SomeSchema.pre('save', saveFunction);
```

unit test
```
var nextSpy = sinon.spy();
saveFunction(nextSpy);
assertTrue(nextSpy.called);
```

**Comment by gaguirre** [comment]
As @jniemin said, I think you can unit test the `pre-save` hook separately.
A test mocking the `save` method,  returning the mongoose instance _as you expect to be modified by the `pre-save` hook_.
Another test just for the hook function.

@newmanw tell me if that's clear enough

**Comment by newmanw** [comment]
@gaguirre yes, thanks.  Only thing that I have trouble with is testing a non exported function.

IE in @jniemin example for that to work I believe you would need to do something like this:

In model.js file
```
var saveFunction = function(next){
  next();
}

SomeSchema.pre('save', saveFunction);

exports.saveFunction = saveFunction;
```

Then in test file

```
var saveFunction = require('./model').saveFunction;
var nextSpy = sinon.spy();
saveFunction(nextSpy);
assertTrue(nextSpy.called);
```

Not really optimal.  I could possibly get around this using rewire.  But no good way IMHO to ask a mongoose model for its 'installed' hooks.

**Comment by gaguirre** [comment]
@newmanw I think it has to be available somewhere in the schema. If you can find where it is then you can test it, and avoid exporting the method in your model.
If you find a better solution post it here! :)

**Comment by jniemin** [comment]
What I sometime do is assign them to object that is clearly marked private. There is not really optimal way of exposing "private" function out of modules. So exposing private object at least makes it explicit that it is a private object. Not perfect anyway.

```

var aFunction = function(){}

var privateAPI = {
 aFunction : aFunction
}

module.exports.privateAPI = privateAPI

```

Then you can call it on tests like

```
var myModule = require('mymodule');
myModule.privateAPI.aFunction();
```

**Comment by vikpe** [comment]
@newmanw I stumbled upon the same issue recently and "solved" it by exposing the schema middleware function and using [lodash.bind()](https://lodash.com/docs/4.17.4#bind) during testing in order to rewrite the `this` context.

I made a [tiny example repo](https://github.com/vikpe/mongoose-middleware-test) with source/tests using this approach.
